data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/BulkRetryStrategy.java

```diff
@@ -252,38 +269,81 @@ private AccumulatingBulkRequest<BulkOperation, BulkRequest> createBulkRequestFor
                     if (!NON_RETRY_STATUS.contains(bulkItemResponse.status())) {
                         requestToReissue.addOperation(request.getOperationAt(index));
                     } else {
-                        // log non-retryable failed request
-                        logFailure.accept(request.getOperationAt(index), new RuntimeException(toSingleLineDisplayString(bulkItemResponse.error())));
+                        nonRetryableFailures.add(convertToDlqObject(request.getOperations().get(index), bulkItemResponse, null));
                         documentErrorsCounter.increment();
                     }
                 } else {
                     sentDocumentsCounter.increment();
                 }
                 index++;
             }
+            logFailure.accept(nonRetryableFailures.build(), null);
             return requestToReissue;
         }
     }
 
     private void handleFailures(final AccumulatingBulkRequest<BulkOperation, BulkRequest> accumulatingBulkRequest, final List<BulkResponseItem> itemResponses) {
         assert accumulatingBulkRequest.getOperationsCount() == itemResponses.size();
+        final ImmutableList.Builder<DlqObject> failures = ImmutableList.builder();
         for (int i = 0; i < itemResponses.size(); i++) {
             final BulkResponseItem bulkItemResponse = itemResponses.get(i);
             final BulkOperation bulkOperation = accumulatingBulkRequest.getOperationAt(i);
             if (bulkItemResponse.error() != null) {
-                final ErrorCause error = bulkItemResponse.error();
-                logFailure.accept(bulkOperation, new RuntimeException(toSingleLineDisplayString(error)));
+                failures.add(convertToDlqObject(bulkOperation, bulkItemResponse, null));
                 documentErrorsCounter.increment();
             } else {
                 sentDocumentsCounter.increment();
             }
         }
+        logFailure.accept(failures.build(), null);
     }
 
     private void handleFailures(final AccumulatingBulkRequest<BulkOperation, BulkRequest> accumulatingBulkRequest, final Throwable failure) {
         documentErrorsCounter.increment(accumulatingBulkRequest.getOperationsCount());
-        for (final BulkOperation bulkOperation: accumulatingBulkRequest.getOperations()) {
-            logFailure.accept(bulkOperation, failure);
+        final ImmutableList.Builder<DlqObject> failures = ImmutableList.builder();
+        final List<BulkOperation> bulkOperations = accumulatingBulkRequest.getOperations();
+        for (int i = 0; i < bulkOperations.size(); i++) {
+            final BulkOperation bulkOperation = bulkOperations.get(i);
+            failures.add(convertToDlqObject(bulkOperation, null, failure));
+        }
+
+        logFailure.accept(failures.build(), failure);
+    }
+
+    private DlqObject convertToDlqObject(final BulkOperation bulkOperation, final BulkResponseItem bulkResponseItem, final Throwable failure) {
```

Comment:
As noted in my comment above, I think you could create a distinct class which handles conversion to a list of DlqObject instances. This will decouple these two classes and separate their concerns.